### PR TITLE
LibWeb+LibGfx: Implement paint styles and all canvas gradients! 

### PR DESCRIPTION
**Pull request by MacDue**
This PR implements paint styles in LibGfx which can be used for filling rects or paths (currently), then uses these to implement _all of_ the HTML5 Canvas gradients :^)

Currently, the paint styles can only be used to fill things, I've not yet got them working for strokes (which is may be a bit trickier).  

(This PR was a little bit of a yak hole as I made the classic blunder of assuming the canvas gradients were the same as the CSS ones...  They're not, but they were all easy, except the canvas radial gradient.. which was a bit of a pain).

**LibWeb:**

[screen-capture - 2023-01-16T234712.210.webm](https://user-images.githubusercontent.com/11597044/212781259-481c1552-b7c2-42e4-9dfe-fafaf9ec9bae.webm)

**Chrome (for context):**

[screen-capture - 2023-01-16T234844.448.webm](https://user-images.githubusercontent.com/11597044/212781423-c52ccbea-fc24-41b8-b04b-f117e912e480.webm)

You can see we render things the same (albeit a little slower :stuck_out_tongue:)

### CODE REVIEW

**Comment by MacDue** [comment]
I wonder how completely different the gradients required for SVG will be :P
(looking at them now looks like they might actually be identical to the canvas ones... so that'd be a win :))

**Comment by MacDue** [comment]
Question: Should I rename `FillStyle` to `PaintStyle` anticipating adding support for strokes to use these too?

**Comment by MacDue** [comment]
`s/FillStyle/PaintStyle/`
 Hopefully a more future proof name :)

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#include <AK/RefPtr.h>
+#include <AK/Vector.h>
+#include <LibGfx/Color.h>
+#include <LibGfx/Forward.h>
+#include <LibGfx/Gradients.h>
+#include <LibGfx/Rect.h>
+
+namespace Gfx {
+
+struct PaintStyle : RefCounted<PaintStyle> {
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:22]: We typically only use `struct` for simple structures with a handful of public members.

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    // Paint styles that have paint time dependent state (e.g. based on the paint size) may find it easier to override paint().
+    // If paint() is overridden sample_color() is unused.
+    virtual void paint(IntRect physical_bounding_box, PaintFunction paint) const
+    {
+        (void)physical_bounding_box;
+        paint([this](IntPoint point) { return sample_color(point); });
+    }
+};
+
+struct SolidPaintStyle : PaintStyle {
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:43]: Maybe `SolidColorPaintStyle` would be even clearer?
Also please mark this `final`.

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+private:
+    Vector<ColorStop, 4> m_color_stops;
+    Optional<float> m_repeat_length;
+};
+
+// These paint styles are based on the CSS gradients. They are relative to the painted
+// shape and support premultiplied alpha.
+
+struct LinearGradientPaintStyle : GradientPaintStyle {
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:89]: Please mark this `final`.

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    LinearGradientPaintStyle(float angle)
+        : m_angle(angle)
+    {
+    }
+
+    float m_angle;
+};
+
+struct ConicGradientPaintStyle : GradientPaintStyle {
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:106]: Please mark this `final`.

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        : m_center(center)
+        , m_start_angle(start_angle)
+    {
+    }
+
+    IntPoint m_center;
+    float m_start_angle;
+};
+
+struct RadialGradientPaintStyle : GradientPaintStyle {
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:125]: Please mark this `final`.

File: Userland/Libraries/LibGfx/GradientPainting.cpp [hunk 6]
```diff
@@ -279,4 +281,164 @@ void RadialGradientPaintStyle::paint(IntRect physical_bounding_box, PaintFunctio
...
+    //    Let y(ω) = (y1-y0)ω + y0
+    //    Let r(ω) = (r1-r0)ω + r0
+    // Let the color at ω be the color at that position on the gradient
+    // (with the colors coming from the interpolation and extrapolation described above).
+    //
+    // 3. For all values of ω where r(ω) > 0, starting with the value of ω nearest to positive infinity and
+    // ending with the value of ω nearest to negative infinity, draw the circumference of the circle with
+    // radius r(ω) at position (x(ω), y(ω)), with the color at ω, but only painting on the parts of the
+    // bitmap that have not yet been painted on by earlier circles in this step for this rendering of the gradient.
+    // ^ Bruh what?
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/GradientPainting.cpp:373]: ?

File: Userland/Libraries/LibGfx/PaintStyle.h [hunk 7]
```diff
@@ -141,4 +141,69 @@ struct RadialGradientPaintStyle : GradientPaintStyle {
...
+    virtual void paint(IntRect physical_bounding_box, PaintFunction paint) const override;
+
+    CanvasConicGradientPaintStyle(FloatPoint center, float start_angle)
+        : m_center(center)
+        , m_start_angle(start_angle)
+    {
+    }
+
+    FloatPoint m_center;
+    float m_start_angle;
```
Comment (awesomekling) [inline Userland/Libraries/LibGfx/PaintStyle.h:183]: Please always explicitly initialize primitive variables and data members.

File: Userland/Libraries/LibGfx/GradientPainting.cpp (hunk 6 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/GradientPainting.cpp:373]: :sweat_smile: I got a little frustrated there... The way the HTML spec defines the radial gradient has no bearing in writing an implementation. I don't know why they don't explain the actual maths, which once you figure it out is quite a bit more obvious. 

I can remove my comment though :smile:

**Pull request merged**